Add cause donation distribution cronjob

ae2079 opened the pull request:
related to #2018


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced an automated job to periodically distribute resources to projects associated with active causes.
  * Added environment settings to enable or schedule the distribution job at a configurable interval (default: every 3 hours).
  * Enabled scheduling of the distribution job during server startup.

* **Chores**
  * Improved logging and error handling for the new distribution process to enhance monitoring and reliability.
  * Refined distribution update logic to correctly accumulate amounts over time.

* **Tests**
  * Added comprehensive tests to verify project eligibility criteria for the distribution job.
  * Enhanced existing tests to ensure accurate accumulation behavior in distribution updates.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ae2079 committed:
Add cause distribution job functionality

- Introduced a new cron job for cause distribution, scheduled to run every 3 hours.
- Implemented the `runCauseDistributionJob` function to process active causes and distribute funds to eligible projects.
- Added `callDistributionService` method in `AgentDistributionService` to handle API calls for fund distribution.
- Updated environment configuration to enable the cause distribution job and set the cron job expression.

ae2079 committed:
Add unit test for testing the cause donation distribution cronjob logics

ae2079 committed:
apply review comments

CarlosQ96 commented on src/services/cronJobs/causeDistributionJob.ts:
```diff
@@ -0,0 +1,183 @@
+
+    // Get all active causes
+    const activeCauses = await Cause.createQueryBuilder('cause')
+      .leftJoinAndSelect('cause.causeProjects', 'causeProjects')
```
Due to a bug, leftJoining doesnt return the cause projects. but you can assign it to the variable of the entity
```
cause.causeProjects = await cause.loadCauseProjects();
```
In case you need. Else can remove the causeProject and its projects since they are extra joins not really fetching data just slowing it down.

CarlosQ96 reviewed (approved):
THanks Ali looks good

ae2079 committed:
Enhance cause project distribution functionality and tests

- Updated `updateCauseProjectDistribution` to accumulate amounts when updating existing records.
- Added validation to ensure cause and project exist before updates.
- Refactored tests to reset cause project state before each test, ensuring accurate accumulation checks.
- Enhanced test cases to verify accumulation of distribution data in both repository and resolver tests.

ae2079 committed:
Merge branch 'staging' of github.com:Giveth/impact-graph into AddDistributionCronjob

ae2079 committed:
Refactor cause distribution job to load cause projects dynamically

- Removed unnecessary left join for cause projects in the query.
- Added logic to load cause projects for each active cause asynchronously.
- Improved overall clarity and efficiency of the cause distribution process.
